python/tvm/autotvm/database.py

```diff
@@ -0,0 +1,187 @@
+# pylint: disable=consider-using-enumerate,invalid-name
+"""
+Database of MeasureInput/MeasureResult pair.
+This can be used for replaying measurement.
+"""
+import os
+
+from .record import encode, decode, measure_str_key
+
+
+class Database(object):
+    """
+    Base class for a record database object.
+    """
+    def load(self, inp, get_all=False):
+        """
+        Load a result based on an input's string key
+
+        Parameters
+        ----------
+        inp: MeasureInput
+            to be translated into key for RedisDB
+        get_all: bool, optional
+            Whether the latest result (or all matching results) should be returned
+
+        Returns
+        -------
+        rec: MeasureResult if previously saved, otherwise None
+        """
+        raise NotImplementedError()
+
+    def save(self, inp, res, extend=False):
+        """
+        Save a result based on an input's string key
+
+        Parameters
+        ----------
+        inp: MeasureInput
+            to be translated into key for RedisDB
+        res: MeasureResult
+            to associate with key
+        extend:
+            Whether to extend existing MeasureResults if they exist
+        """
+        raise NotImplementedError()
+
+    def filter_inputs(self, measure_inputs, retry=False):
+        """
+        Filter a measure_inputs batch based on saved db results
+
+        Parameters
+        ----------
+        measure_inputs: Array of MeasureInput
+            measure_inputs as expected in measure_batch
+        retry: bool
+            whether to retry if the saved result is a failure
+
+        Returns
+        -------
+        partial_results: Array of MeasureResult
+            a full list of result, where None denotes no corresponding saved result
+        unsaved: Array of MeasureInput
+            a list that only contains unsaved inputs
+        """
+        partial_results = [None] * len(measure_inputs)
+        unsaved = list()
+        for i in range(len(measure_inputs)):
+            inp = measure_inputs[i]
+            res = self.load(inp)
+            skip = (res is not None and
+                    (not retry or (retry and res.error_no == 0)))
+            if skip:
+                partial_results[i] = res
+            else:
+                unsaved.append(inp)
+        return partial_results, unsaved
+
+
+def unpack_existing(current):
+    """
+    Unpack existing MeasureResults (in str format)
+
+    Parameters
+    ----------
+    current: str
+        The current list of MeasureResults (in str format)
+    """
+    # pylint: disable = eval-used
+    return eval(current)
```

Comment:
eval is dangerous and should be avoided, if it is json, use json.loads,

Reply:
This might have just been a quick and dirty way to get the encoding to work, since redis expects strings (and returns back bytes)? Might be able to try record encode -> str encode -> set, and get -> str decode -> record decode.